Add weight of evidence encoder

tlienart opened the pull request:
* adds `WOEEncoder` to the encoders described here https://www.listendata.com/2015/03/weight-of-evidence-woe-and-information.html
* WOE is a supervised encoder in the case of *binary* output which can encode
  * categorical data
  * numerical data (after binning)
* The current implementation assumes that the data has previously gone through an imputer and so assumes that there is no missing values (either in `X` or `y`). 

**Notes**:

* adds a `titanic.csv` dataset for testing, the testing is done against sklearn-contrib (https://contrib.scikit-learn.org/category_encoders/woe.html)
* it's a supervised encoder expecting binary target; this doesn't work well with ScikitLearn's automatic `check_estimator` which tries to fit the estimator with Boston (3 classes). **Please advise**.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

cc @jeanfad

### Edits

* f30f650 fixes
  * a line too long issue (reference to sklearn)
  * an except with an error variable that was unused
  * a naming issue (variable named `iter`, pylint doesn't like that.

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 69e6b2f9008aaea04bfe7533787004aba4814e90
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=61a23c5b-6c48-4908-9b03-4d572bbfe4f6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 165902d1497e5df89d739c4518852d7794e71ca8
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=1593f04c-affb-4eff-bc27-372bae8454bb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: f30f6500fe7910b48fd67bb976481988e6680788
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4b70f88e-5ff9-45d7-88ad-d06cceecbfce%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented on test/test_common.py:
```diff
@@ -47,6 +48,7 @@
         RobustMissingIndicator(),
         RobustStandardScaler(),
         ThresholdOneHotEncoder(),
+        # WOEEncoder(),
```
Please uncomment this

tlienart replied:
this won't work as per PR comment, please advise

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+
+    Attributes
+    ----------
+    _beta: float
```
Documentation shows `_beta` but code shows `beta_`. Documentation needs to be updated

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+        self.alpha = alpha
+        self.laplace = laplace
+
+        # fail early for arguments that don't meet the guidelines
+        if binning:
+            assert binning in ("uniform", "quantile", "kmeans"), \
+                WOEAsserts.BINNING
+            assert n_bins > 2, WOEAsserts.NBINS
+        assert alpha >= 0, WOEAsserts.ALPHA
```
I believe this will fail `check_estimator`, there's a requirement that all input validation be done in `fit` rather than the initialization.

tlienart replied:
I'll try that

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+
+    Parameters
+    ----------
+    feature_indices: list of integer indices (default=None)
+        Indices of the features to encode with WoE. If set to `None`, all
+        features will be encoded.
```
This is unnecessary, the column selection functionality should be used by the [`ColumnTransformer`](https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html)

tlienart replied:
@jeanfad, I don't mind dropping this.

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+        except TypeError:
+            pass
+
+    def _feature_iterator(self, X):
```
This can most likely be removed if the `feature_indices` parameter is removed

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+        -------
+        self: WOEEncoder.
+        """
+        X, y = check_X_y(X, y)
```
I recommend using `sklearn.utils.check_array`, especially if the X and y inputs are "array-like". There should also be an option for check_array to not accept missing values, eliminating the need for L817

tlienart replied:
https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_X_y.html?highlight=check_x_y#sklearn.utils.check_X_y

* effectively calls `check_array` (on both `X` and `y`) (and you're right that it checks for missing so I can scrap that)

wiltonwu commented on test/test_preprocessing_encoders.py:
```diff
@@ -254,3 +258,66 @@ def test_robust_ordinal_encoding_inverse_transform_floatkeys():
+    uv = np.sort(np.abs(np.unique(xe)))
+    assert np.allclose(uv, SK_CONTRIB_0)
+
+    # Laplace smoothing
```
Use `pytest.mark.parametrize` instead of having two test cases in one test

wiltonwu commented on test/test_preprocessing_encoders.py:
```diff
@@ -254,3 +258,66 @@ def test_robust_ordinal_encoding_inverse_transform_floatkeys():
     assert all([x is None for x in reverse[-1]])
+
+
+titanic = pd.read_csv("test/data/csv/titanic.csv")
```
Let's use `sagemaker_sklearn_extension.externals.read_csv_data` instead of pandas

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+    _count_y: list of size 2
+        Number of observations for each of the two target classes.
+
+    _mask_y_0: array (n_samples,)
+        Mask of observations for which the target is the first one.
+
+    _mask_y_1: array (n_samples,)
+        Mask of observations for which the target is the second one.
```
Do these mask values need to be stored as attributes? I don't see them used in the `transform` step

tlienart commented on setup.py:
```diff
@@ -36,5 +36,5 @@ def read_version():
     keywords="ML Amazon AWS AI SKLearn Scikit-Learn",
     classifiers=["Development Status :: 4 - Beta", "License :: OSI Approved :: Apache Software License"],
-    extras_require={"test": ["tox", "tox-conda", "pytest", "coverage"]},
+    extras_require={"test": ["tox", "tox-conda", "pytest", "coverage", "pandas"]},
```
this is for the tests, why does it matter? it won't be installed by the user. And it allows to test more thoroughly with categorical features and missing values.

tlienart commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+        features will be encoded.
+
+    binning: {'uniform', 'quantile', 'kmeans', None}, default=None
+        What binning method to apply if not set to None. See [2].
```
No binning is applied. ok to briefly clarify the docs.

tlienart commented:
Thanks a lot for the review, I'll fix everything I can asap

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: f30f6500fe7910b48fd67bb976481988e6680788
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=4f046f4e-1342-4247-8744-69bc1fc2cbf2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 19c6c4e4d7920748b647eda42274cf8562ff96e5
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=fe762c5c-da30-43ff-8405-7467836e4d33%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 6f8ac9739768701a6576c344ddd48c4082c241c1
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=bf7a859e-a49c-44ad-a686-16edde5a2a37%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

tlienart commented:
Right I think I addressed all your comments & extended the spirit of some (e.g. removed all attributes unused in transform).

The only remaining point as mentioned in the comment is that you seem to want the `check_estimator` to be applied to WoE, I may be missing something but the check_estimator uses boston which has three classes and therefore will always fail with WoE. My suggestion is to ignore it and, if there is concern that we may be missing something, suggest the explicit addition of extra tests for WoE that are related to dimensions etc.

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 6c11ab6f98d134714612c87287e990ea82c26e34
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=895fb3e7-cdb8-424a-9006-039ff6ec4289%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 3c951a5d5fda81b154ca5aabbcd702258483c75e
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=0f7bb2d8-52e0-41ab-9a57-41d65c45748c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 269b4b6cdb6a057ad6c3cd9d8de3a95e088dddaa
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b2e86613-0e06-4c45-94c9-1429a92dbfa5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

tlienart commented:
gentle bump, thanks!

wiltonwu commented:
> Right I think I addressed all your comments & extended the spirit of some (e.g. removed all attributes unused in transform). Pandas is removed.
> 
> The only remaining point as mentioned in the comments is that you seem to want the `check_estimator` to be applied to WoE, I may be missing something but the check_estimator uses boston which has three classes and therefore will always fail with WoE. My suggestion is to ignore it and, if there is concern that we may be missing something, suggest the explicit addition of extra tests for WoE that are related to dimensions etc.

I think the build logs still show some flake8 and black linting tests failing. Those should be straightforward to fix, please let me know if you need help.

For the `check_estimator` test, I think you can add a tag for your new estimator to mark it as only compatible with binary classification datasets. It involves override the `_get_tags()` method. See https://scikit-learn.org/stable/developers/develop.html#estimator-tags

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 4034f8373fa7c4f2cc07fe442ea74bda15081acf
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=2f02e852-bfd8-4922-960a-38ed9233eaba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

tlienart commented:
* fixed what was requested by the flake8 ; it's unclear to me what the black linting wants (concrete help on specific lines to adjust appreciated)
* I ended up implicitly turning off the check_estimator by setting `'X_types': ['categorical']` because it tests by default with continuous values and this causes issues in the later checks at the transform stage (in testing whether what gets encoded has new categories); I did fix a few minor things based on the earlier tests though.

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: aad5e404980b0e3ef9c832d38b505203cec6570b
* Result: FAILED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=244fd256-61bf-4fb4-ae65-6578bbe6772c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: 4e04fada960188315973e2c4d54c17cb7ffc2ec0
* Result: SUCCEEDED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=6bc0fe88-d693-418b-88b3-6b4d8e21b479%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,244 @@ def inverse_transform(self, X):
+    """
+
+    def __init__(self, feature_indices=None, binning=None, n_bins=10, alpha=0.5, laplace=False):
+        self.feature_indices = feature_indices
```
Can this be removed now? I don't see it being used anywhere?

wiltonwu commented on src/sagemaker_sklearn_extension/preprocessing/encoders.py:
```diff
@@ -620,3 +621,244 @@ def inverse_transform(self, X):
+    BINARY = "Weight-of-Evidence encoder is only supported for binary targets."
+    BINNING = "Binning strategy must be in {'uniform', 'quantile', 'kmeans'}."
+    NBINS = "Number of bins must be larger than 2."
+    DIMS_TRANSFORM = "The number of columns of `X` doesn't match the encoder."
```
This can also be removed

wiltonwu commented:
### AWS CodeBuild CI Report

* CodeBuild project: sagemaker-sklearn-extension-pr
* Commit ID: c3988864e44ab9d38173565b735e7d5419da9922
* Result: SUCCEEDED
* [Build Logs](https://cdvka32a8a.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=30bab70a-f4aa-4570-a8a5-e1c7305c07ae%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*
